### PR TITLE
fuzz: Reset addrman when consistency check fails

### LINKED ISSUE #22503: Addrman check failure info.nLastSuccess is 0 even though info.fInTried
Using #22493 + #22500 gives the failure mentioned in the subject line

[crash-2e85f5a8d3623503ed9c668d02dc65c5bb573df1.log](https://github.com/bitcoin/bitcoin/files/6846105/crash-2e85f5a8d3623503ed9c668d02dc65c5bb573df1.log)

### LINKED ISSUE #22504: Addrman check failure info.nRefCount is non-zero even though !info.fInTried
Using #22493 + #22500 gives the failure mentioned in the subject line

[crash-6ddb0b69a79d653752c09a9000e9006f43de75fc.log](https://github.com/bitcoin/bitcoin/files/6846129/crash-6ddb0b69a79d653752c09a9000e9006f43de75fc.log)

### LINKED ISSUE #22519: Addrman check failure: info.nLastSuccess is negative
Using #22493 + #22500 gives the failure mentioned in the subject line

[crash-f11ac19bfb2bcd68149035337afede0e56848596.log](https://github.com/bitcoin/bitcoin/files/6856720/crash-f11ac19bfb2bcd68149035337afede0e56848596.log)

### LINKED ISSUE #22931: Signed Integer-overflow in CAddrInfo::IsTerrible
This may happen when peers.dat is deserialized from disk.

See:

* #22925
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=38301

### PR DESCRIPTION
**Pull request by maflcko**
It doesn't really make sense to continue when the deserialized addrman isn't consistent in the `addrman` test.

For example, if `nLastSuccess` is negative, it would  later result in integer overflows. Thus, this patch fixes #22931.

Also,
Fixes #22503
Fixes #22504
Fixes #22519

### CODE REVIEW

**Review by amitiuttarwar — COMMENTED** [review]
concept ACK

**Comment by mzumsande** [comment]
Concept ACK. 

I'd understand if it's beyond the scope of this PR, but I think that a second `ForceCheckAddrman()`  at the end of the  `addrman` fuzz target (which isn't caught but can actually lead to crashes) would be really helpful - probably better than activating  `consistency_check_ratio` during fuzzing which could make the fuzzer slow if called in too many steps.

**Comment by MarcoFalke** [comment]
> a second ForceCheckAddrman() at the end of the addrman fuzz target

Good idea. Wanted to do it in a follow-up, but it is just a one-line fix, so I added the commit here.

File: src/addrman.h [hunk 1]
```diff
@@ -391,20 +391,11 @@ class CAddrMan
...
-    {
-        AssertLockHeld(cs);
-
-        const int err = Check_();
-        if (err) {
-            LogPrintf("ADDRMAN CONSISTENCY CHECK FAILED!!! err=%i\n", err);
-            assert(false);
-        }
-    }
+    //! Consistency check, taking into account m_consistency_check_ratio.
```
Comment (vasild) [inline src/addrman.h:394]:
```suggestion
    //! Consistency check, taking into account m_consistency_check_ratio. Will abort(2) if an inconsistency is detected.
```

File: src/addrman.h [hunk 2]
```diff
@@ -391,20 +391,11 @@ class CAddrMan
...
+    void Check() const EXCLUSIVE_LOCKS_REQUIRED(cs);
 
-    //! Perform consistency check. Returns an error code or zero.
-    int Check_() const EXCLUSIVE_LOCKS_REQUIRED(cs);
+    //! Perform consistency check, regardless of m_consistency_check_ratio. Returns an error code or zero.
```
Comment (vasild) [inline src/addrman.h:397]:
```suggestion
    //! Perform consistency check, regardless of `m_consistency_check_ratio`.
    //! @returns an error code or zero.
```

File: src/test/fuzz/addrman.cpp [hunk 3]
```diff
@@ -238,6 +241,9 @@ FUZZ_TARGET_INIT(addrman, initialize_addrman)
         ds.SetVersion(ser_version);
         try {
             ds >> *addr_man_ptr;
+            if (0 != WITH_LOCK(addr_man_ptr->cs, return addr_man_ptr->ForceCheckAddrman())) {
+                throw std::ios_base::failure{"Reset addrman after consistency check failed"};
+            }
```
Comment (vasild) [inline src/test/fuzz/addrman.cpp:246]: What about moving this unconditional check inside, at the end of `CAddrMan::Unserialize()`? We never want to continue if this check fails after unserialize and it is not too expensive to check once per unserialize.

Right now we have the conditional `Check()` at the end of `Unserialize()`.

Comment (MarcoFalke) [reply src/test/fuzz/addrman.cpp:246]: This is what #22734 is doing. Happy to close this fuzz-only pull and rebase the other.

Comment (vasild) [reply src/test/fuzz/addrman.cpp:246]: Up to you (and other reviewers). I "voted" for the other PR :)

**Pull request closed without merging**